File: src/main/java/com/zyuc/iot/java/kafka/KafkaOffsetTool.java
package com.zyuc.iot.java.kafka;


import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import kafka.api.PartitionOffsetRequestInfo;
import kafka.cluster.Broker;
import kafka.cluster.BrokerEndPoint;
import kafka.common.TopicAndPartition;
import kafka.javaapi.*;
import kafka.javaapi.consumer.SimpleConsumer;
import org.apache.spark.streaming.kafka.KafkaCluster;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * KafkaOffsetTool
 *
 * @author FREY
 * @date 2016/4/11
 */
public class KafkaOffsetTool {

    private static KafkaOffsetTool instance;
    final int TIMEOUT = 100000;
    final int BUFFERSIZE = 64 * 1024;

    private KafkaOffsetTool() {
    }

    public static synchronized KafkaOffsetTool getInstance() {
        if (instance == null) {
            instance = new KafkaOffsetTool();
        }
        return instance;
    }


    public Map<TopicAndPartition, Long> getLastOffset(String brokerList, List<String> topics,
                                                      String groupId) {

        Map<TopicAndPartition, Long> topicAndPartitionLongMap = Maps.newHashMap();

        Map<TopicAndPartition, BrokerEndPoint> topicAndPartitionBrokerMap =
                KafkaOffsetTool.getInstance().findLeader(brokerList, topics);

        for (Map.Entry<TopicAndPartition, BrokerEndPoint> topicAndPartitionBrokerEntry : topicAndPartitionBrokerMap
                .entrySet()) {
            // get leader broker
            BrokerEndPoint leaderBroker = topicAndPartitionBrokerEntry.getValue();

            SimpleConsumer simpleConsumer = new SimpleConsumer(leaderBroker.host(), leaderBroker.port(),
                    TIMEOUT, BUFFERSIZE, groupId);

            long readOffset = getTopicAndPartitionLastOffset(simpleConsumer,
                    topicAndPartitionBrokerEntry.getKey(), groupId);

            topicAndPartitionLongMap.put(topicAndPartitionBrokerEntry.getKey(), readOffset);

        }

        return topicAndPartitionLongMap;

    }

    /**
     *
     * @param brokerList
     * @param topics
     * @param groupId
     * @return
     */
    public Map<TopicAndPartition, Long> getEarliestOffset(String brokerList, List<String> topics,
                                                          String groupId) {

        Map<TopicAndPartition, Long> topicAndPartitionLongMap = Maps.newHashMap();

        Map<TopicAndPartition, BrokerEndPoint> topicAndPartitionBrokerMap =
                KafkaOffsetTool.getInstance().findLeader(brokerList, topics);

        for (Map.Entry<TopicAndPartition, BrokerEndPoint> topicAndPartitionBrokerEntry : topicAndPartitionBrokerMap
                .entrySet()) {
            // get leader broker
            BrokerEndPoint leaderBroker = topicAndPartitionBrokerEntry.getValue();

            SimpleConsumer simpleConsumer = new SimpleConsumer(leaderBroker.host(), leaderBroker.port(),
                    TIMEOUT, BUFFERSIZE, groupId);

            long readOffset = getTopicAndPartitionEarliestOffset(simpleConsumer,
                    topicAndPartitionBrokerEntry.getKey(), groupId);

            topicAndPartitionLongMap.put(topicAndPartitionBrokerEntry.getKey(), readOffset);

        }

        return topicAndPartitionLongMap;

    }

    /**
     * 得到所有的 TopicAndPartition
     *
     * @param brokerList
     * @param topics
     * @return topicAndPartitions
     */
    private Map<TopicAndPartition, BrokerEndPoint> findLeader(String brokerList, List<String> topics) {
        // get broker's url array
        String[] brokerUrlArray = getBorkerUrlFromBrokerList(brokerList);
        // get broker's port map
        Map<String, Integer> brokerPortMap = getPortFromBrokerList(brokerList);

        // create array list of TopicAndPartition
        Map<TopicAndPartition, BrokerEndPoint> topicAndPartitionBrokerMap = Maps.newHashMap();

        for (String broker : brokerUrlArray) {

            SimpleConsumer consumer = null;
            try {
                // new instance of simple Consumer
                consumer = new SimpleConsumer(broker, brokerPortMap.get(broker), TIMEOUT, BUFFERSIZE,
                        "leaderLookup" + new Date().getTime());

                TopicMetadataRequest req = new TopicMetadataRequest(topics);

                TopicMetadataResponse resp = consumer.send(req);

                List<TopicMetadata> metaData = resp.topicsMetadata();

                for (TopicMetadata item : metaData) {
                    for (PartitionMetadata part : item.partitionsMetadata()) {
                        TopicAndPartition topicAndPartition =
                                new TopicAndPartition(item.topic(), part.partitionId());
                        topicAndPartitionBrokerMap.put(topicAndPartition, part.leader());
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                if (consumer != null)
                    consumer.close();
            }
        }
        return topicAndPartitionBrokerMap;
    }

    /**
     * get last offset
     * @param consumer
     * @param topicAndPartition
     * @param clientName
     * @return
     */
    private long getTopicAndPartitionLastOffset(SimpleConsumer consumer,
                                                TopicAndPartition topicAndPartition, String clientName) {
        Map<TopicAndPartition, PartitionOffsetRequestInfo> requestInfo =
                new HashMap<TopicAndPartition, PartitionOffsetRequestInfo>();

        requestInfo.put(topicAndPartition, new PartitionOffsetRequestInfo(
                kafka.api.OffsetRequest.LatestTime(), 1));

        OffsetRequest request = new OffsetRequest(
                requestInfo, kafka.api.OffsetRequest.CurrentVersion(),
                clientName);

        OffsetResponse response = consumer.getOffsetsBefore(request);

        if (response.hasError()) {
            System.out
                    .println("Error fetching data Offset Data the Broker. Reason: "
                            + response.errorCode(topicAndPartition.topic(), topicAndPartition.partition()));
            return 0;
        }
        long[] offsets = response.offsets(topicAndPartition.topic(), topicAndPartition.partition());
        return offsets[0];
    }

    /**
     * get earliest offset
     * @param consumer
     * @param topicAndPartition
     * @param clientName
     * @return
     */
    private long getTopicAndPartitionEarliestOffset(SimpleConsumer consumer,
                                                    TopicAndPartition topicAndPartition, String clientName) {
        Map<TopicAndPartition, PartitionOffsetRequestInfo> requestInfo =
                new HashMap<TopicAndPartition, PartitionOffsetRequestInfo>();

        requestInfo.put(topicAndPartition, new PartitionOffsetRequestInfo(
                kafka.api.OffsetRequest.EarliestTime(), 1));

        OffsetRequest request = new OffsetRequest(
                requestInfo, kafka.api.OffsetRequest.CurrentVersion(),
                clientName);

        OffsetResponse response = consumer.getOffsetsBefore(request);

        if (response.hasError()) {
            System.out
                    .println("Error fetching data Offset Data the Broker. Reason: "
                            + response.errorCode(topicAndPartition.topic(), topicAndPartition.partition()));
            return 0;
        }
        long[] offsets = response.offsets(topicAndPartition.topic(), topicAndPartition.partition());
        return offsets[0];
    }
    /**
     * 得到所有的broker url
     *
     * @param brokerlist
     * @return
     */
    private String[] getBorkerUrlFromBrokerList(String brokerlist) {
        String[] brokers = brokerlist.split(",");
        for (int i = 0; i < brokers.length; i++) {
            brokers[i] = brokers[i].split(":")[0];
        }
        return brokers;
    }

    /**
     * 得到broker url 与 其port 的映射关系
     *
     * @param brokerlist
     * @return
     */
    private Map<String, Integer> getPortFromBrokerList(String brokerlist) {
        Map<String, Integer> map = new HashMap<String, Integer>();
        String[] brokers = brokerlist.split(",");
        for (String item : brokers) {
            String[] itemArr = item.split(":");
            if (itemArr.length > 1) {
                map.put(itemArr[0], Integer.parseInt(itemArr[1]));
            }
        }
        return map;
    }

    public static void main(String[] args) {
        List<String> topics = Lists.newArrayList();
        topics.add("test");


//    topics.add("bugfix");
        Map<TopicAndPartition, Long> topicAndPartitionLongMap =
                KafkaOffsetTool.getInstance().getEarliestOffset("192.168.5:9091", topics,
                        "com.frey.group");

        for (Map.Entry<TopicAndPartition, Long> entry : topicAndPartitionLongMap.entrySet()) {
            System.out.println(entry.getKey().topic() + "-"+ entry.getKey().partition() + ":" + entry.getValue());
        }
    }
}

File: src/test/scala/test/KafkaOffsetTool.scala
package test

/**
  * Created by zhoucw on 17-9-25.
  */

import com.google.common.collect.Lists
import com.google.common.collect.Maps
import kafka.api.PartitionOffsetRequestInfo
import kafka.cluster.BrokerEndPoint
import kafka.common.TopicAndPartition
import kafka.javaapi._
import kafka.javaapi.consumer.SimpleConsumer
import java.util.Date
import java.util

import com.zyuc.iot.java.kafka.RadiusReceive.prop
import org.apache.spark.streaming.kafka._


/**
  * KafkaOffsetTool
  *
  */
object KafkaOffsetTool {

  val kafkaParams = Map[String, String]("metadata.broker.list" -> "")



   var instance:KafkaOffsetTool = null
  def getInstance: KafkaOffsetTool = {
    if (instance == null) instance = new KafkaOffsetTool
    instance
  }

  def main(args: Array[String]): Unit = {
    val topics = Lists.newArrayList[String]
    topics.add("haradius_out")
    //    topics.add("bugfix");
    val topicAndPartitionLongMap = KafkaOffsetTool.getInstance.getEarliestOffset("10.37.7.139:9092,10.37.7.140:9092,10.37.7.141:9092", topics, "haradiusToHive")
    val topicAndPartitionLongMap1 = KafkaOffsetTool.getInstance.getLastOffset("10.37.7.139:9092,10.37.7.140:9092,10.37.7.141:9092", topics, "haradiusToHive")
    import scala.collection.JavaConversions._
    for (entry <- topicAndPartitionLongMap.entrySet) {
      System.out.println(entry.getKey.topic + "-" + entry.getKey.partition + ":" + entry.getValue)
    }

    import kafka.common.TopicAndPartition
    val lastestTopicAndPartitionLongMap = KafkaOffsetTool.getInstance.getLastOffset("10.37.7.139:9092,10.37.7.140:9092,10.37.7.141:9092", topics, "haradiusToHive")

    // earliest offsets
    val earliestTopicAndPartitionLongMap = KafkaOffsetTool.getInstance.getEarliestOffset("10.37.7.139:9092,10.37.7.140:9092,10.37.7.141:9092", topics, "haradiusToHive")


    import scala.collection.JavaConversions._


  }
}

class KafkaOffsetTool {
  final val TIMEOUT = 100000
  final val BUFFERSIZE = 64 * 1024

  def getLastOffset(brokerList: String, topics: util.List[String], groupId: String): util.Map[TopicAndPartition, Long] = {
    val topicAndPartitionLongMap = Maps.newHashMap[TopicAndPartition, Long]
    val topicAndPartitionBrokerMap = findLeader(brokerList, topics)
    import scala.collection.JavaConversions._
    for (topicAndPartitionBrokerEntry <- topicAndPartitionBrokerMap.entrySet) { // get leader broker
      val leaderBroker = topicAndPartitionBrokerEntry.getValue
      val simpleConsumer = new SimpleConsumer(leaderBroker.host, leaderBroker.port, TIMEOUT, BUFFERSIZE, groupId)
      val readOffset = getTopicAndPartitionLastOffset(simpleConsumer, topicAndPartitionBrokerEntry.getKey, groupId)
      topicAndPartitionLongMap.put(topicAndPartitionBrokerEntry.getKey, readOffset)
    }
    topicAndPartitionLongMap
  }

  /**
    *
    * @param brokerList
    * @param topics
    * @param groupId
    * @return
    */
  def getEarliestOffset(brokerList: String, topics: util.List[String], groupId: String): util.Map[TopicAndPartition, Long] = {
    val topicAndPartitionLongMap = Maps.newHashMap[TopicAndPartition, Long]
    val topicAndPartitionBrokerMap = findLeader(brokerList, topics)
    import scala.collection.JavaConversions._
    for (topicAndPartitionBrokerEntry <- topicAndPartitionBrokerMap.entrySet) {
      val leaderBroker = topicAndPartitionBrokerEntry.getValue
      val simpleConsumer = new SimpleConsumer(leaderBroker.host, leaderBroker.port, TIMEOUT, BUFFERSIZE, groupId)
      val readOffset = getTopicAndPartitionEarliestOffset(simpleConsumer, topicAndPartitionBrokerEntry.getKey, groupId)
      topicAndPartitionLongMap.put(topicAndPartitionBrokerEntry.getKey, readOffset)
    }
    topicAndPartitionLongMap
  }

  /**
    * 得到所有的 TopicAndPartition
    *
    * @param brokerList
    * @param topics
    * @return topicAndPartitions
    */
   def findLeader(brokerList: String, topics: util.List[String]) = { // get broker's url array
    val brokerUrlArray = getBorkerUrlFromBrokerList(brokerList)
    // get broker's port map
    val brokerPortMap = getPortFromBrokerList(brokerList)
    // create array list of TopicAndPartition
    val topicAndPartitionBrokerMap = Maps.newHashMap[TopicAndPartition, BrokerEndPoint]
    for (broker <- brokerUrlArray) {
      var consumer:SimpleConsumer = null
      try { // new instance of simple Consumer
        consumer = new SimpleConsumer(broker, brokerPortMap.get(broker), TIMEOUT, BUFFERSIZE, "leaderLookup" + new Date().getTime)
        val req = new TopicMetadataRequest(topics)
        val resp = consumer.send(req)
        val metaData = resp.topicsMetadata
        import scala.collection.JavaConversions._
        for (item <- metaData) {
          import scala.collection.JavaConversions._
          for (part <- item.partitionsMetadata) {
            val topicAndPartition = new TopicAndPartition(item.topic, part.partitionId)
            topicAndPartitionBrokerMap.put(topicAndPartition, part.leader)
          }
        }
      } catch {
        case e: Exception =>
          e.printStackTrace()
      } finally if (consumer != null) consumer.close()
    }
    topicAndPartitionBrokerMap
  }

  /**
    * get last offset
    *
    * @param consumer
    * @param topicAndPartition
    * @param clientName
    * @return
    */
   def getTopicAndPartitionLastOffset(consumer: SimpleConsumer, topicAndPartition: TopicAndPartition, clientName: String):Long = {
    val requestInfo = new util.HashMap[TopicAndPartition, PartitionOffsetRequestInfo]
    requestInfo.put(topicAndPartition, new PartitionOffsetRequestInfo(kafka.api.OffsetRequest.LatestTime, 1))
    val request = new OffsetRequest(requestInfo, kafka.api.OffsetRequest.CurrentVersion, clientName)
    val response = consumer.getOffsetsBefore(request)
    if (response.hasError) {
      System.out.println("Error fetching data Offset Data the Broker. Reason: " + response.errorCode(topicAndPartition.topic, topicAndPartition.partition))
      return 0
    }
    val offsets = response.offsets(topicAndPartition.topic, topicAndPartition.partition)
    offsets(0)
  }

  /**
    * get earliest offset
    *
    * @param consumer
    * @param topicAndPartition
    * @param clientName
    * @return
    */
   def getTopicAndPartitionEarliestOffset(consumer: SimpleConsumer, topicAndPartition: TopicAndPartition, clientName: String) :Long = {
    val requestInfo = new util.HashMap[TopicAndPartition, PartitionOffsetRequestInfo]
    requestInfo.put(topicAndPartition, new PartitionOffsetRequestInfo(kafka.api.OffsetRequest.EarliestTime, 1))
    val request = new OffsetRequest(requestInfo, kafka.api.OffsetRequest.CurrentVersion, clientName)
    val response = consumer.getOffsetsBefore(request)
    if (response.hasError) {
      System.out.println("Error fetching data Offset Data the Broker. Reason: " + response.errorCode(topicAndPartition.topic, topicAndPartition.partition))
      return 0
    }
    val offsets = response.offsets(topicAndPartition.topic, topicAndPartition.partition)
    offsets(0)
  }

  /**
    * 得到所有的broker url
    *
    * @param brokerlist
    * @return
    */
   def getBorkerUrlFromBrokerList(brokerlist: String) = {
    val brokers = brokerlist.split(",")
    var i = 0
    while ( {
      i < brokers.length
    }) {
      brokers(i) = brokers(i).split(":")(0)

      {
        i += 1; i - 1
      }
    }
    brokers
  }

  /**
    * 得到broker url 与 其port 的映射关系
    *
    * @param brokerlist
    * @return
    */
   def getPortFromBrokerList(brokerlist: String) = {
    val map = new util.HashMap[String, Integer]
    val brokers = brokerlist.split(",")
    for (item <- brokers) {
      val itemArr = item.split(":")
      if (itemArr.length > 1) map.put(itemArr(0), itemArr(1).toInt)
    }
    map
  }
}